core/src/main/scala/com.snowplowanalytics.stream/loader/sinks/KinesisSink.scala

```diff
@@ -78,40 +80,45 @@ class KinesisSink(conf: KinesisSinkConfig) extends ISink {
       case rnfe: ResourceNotFoundException => false
     }
 
-  private def put(name: String, data: ByteBuffer, key: String): PutRecordResult = {
-    val putRecordRequest = {
-      val p = new PutRecordRequest()
-      p.setStreamName(name)
-      p.setData(data)
-      p.setPartitionKey(key)
-      p
+  private def put(name: String, keyedData: List[KeyedData]): PutRecordsResult = {
+    val prres = keyedData.map { case (key, data) =>
+      new PutRecordsRequestEntry()
+        .withPartitionKey(key)
+        .withData(ByteBuffer.wrap(data))
     }
-    client.putRecord(putRecordRequest)
+    val putRecordsRequest =
+      new PutRecordsRequest()
+        .withStreamName(name)
+        .withRecords(prres.asJava)
+    client.putRecords(putRecordsRequest)
   }
 
   /**
-   * Write a record to the Kinesis stream
+   * Write records to the Kinesis stream
    *
-   * @param output The string record to write
-   * @param key A hash of the key determines to which shard the
-   *            record is assigned. Defaults to a random string.
+   * @param outputs The string records to write
    * @param good Unused parameter which exists to extend ISink
    */
-  def store(output: String, key: Option[String], good: Boolean): Unit =
-    Try {
-      put(
-        conf.streamName,
-        ByteBuffer.wrap(output.getBytes(UTF_8)),
-        key.getOrElse(Random.nextInt.toString)
-      )
-    } match {
-      case Success(result) =>
-        log.info("Writing successful")
-        log.info(s"  + ShardId: ${result.getShardId}")
-        log.info(s"  + SequenceNumber: ${result.getSequenceNumber}")
-      case Failure(f) =>
-        log.error("Writing failed")
-        log.error("  + " + f.getMessage)
+  def store(outputs: List[String], good: Boolean): Unit =
+    groupOutputs(conf) {
+      outputs.map(s => Random.nextInt.toString -> s.getBytes(UTF_8))
+    }.foreach { keyedData =>
+      Try {
+        put(
+          conf.streamName,
+          keyedData
+        )
+      } match {
```

Comment:
It would be great to add some retry mechanism for the writing

Reply:
I was expecting this criticism!!  You're completely right, we need it.

I decided not to include it in this release, because I'm trying to make this release as light as possible on new features.  I really want to get these changes rolled out to production pipelines, to discover if it improves the mysterious stability issues.  I'm reluctant to add in a new "feature" like retries if it could cloud visibility on whether these improvements are working.

But it's a weak argument, and weakly held.  If you strongly think we should add retries now then I'm happy to add it.

Reply:
I understand your reasoning. I'm fine with omitting the retries for now, as they weren't here in the first place, we haven't removed a feature.